File: src/python/underchex/openingbook.py
"""
Underchex Opening Book Module

Provides opening book functionality for the AI:
- Store and lookup opening book positions
- Generate opening book from self-play games
- Probabilistic move selection based on win rates

Opening book format:
- Positions are keyed by board hash (simple string hash)
- Each position stores a list of candidate moves with statistics
- Moves are selected probabilistically based on win rate and play count

Signed-by: agent #31 claude-sonnet-4 via opencode 20260122T08:36:49
"""
from __future__ import annotations

import json
import math
import random
from dataclasses import dataclass, field, asdict
from typing import Optional, Literal, Any
from datetime import datetime

from .types import (
    HexCoord,
    Piece,
    PieceType,
    Color,
    BoardState,
    Move,
    coord_to_string,
    opposite_color,
)

from .moves import (
    generate_all_legal_moves,
    apply_move,
)

from .ai import generate_board_hash

# ============================================================================
# Opening Book Types
# ============================================================================


@dataclass
class BookMoveStats:
    """Statistics for a single move in the opening book."""
    from_q: int
    from_r: int
    to_q: int
    to_r: int
    play_count: int = 0
    wins: int = 0
    draws: int = 0
    avg_score: float = 0.0
    promotion: Optional[PieceType] = None


@dataclass
class BookEntry:
    """Entry in the opening book for a single position."""
    hash_key: str
    moves: list[BookMoveStats] = field(default_factory=list)
    total_visits: int = 0


@dataclass
class BookMetadata:
    """Metadata for the opening book."""
    created_at: str = ""
    games_count: int = 0
    max_depth: int = 0


@dataclass
class OpeningBook:
    """The full opening book structure."""
    entries: dict[str, BookEntry] = field(default_factory=dict)
    metadata: BookMetadata = field(default_factory=BookMetadata)


@dataclass
class BookLookupOptions:
    """Options for opening book lookup."""
    min_play_count: int = 3
    temperature: float = 1.0
    use_win_rate_weight: bool = True
    seed: Optional[int] = None


@dataclass
class BookLookupResult:
    """Result of opening book lookup."""
    move: Optional[Move]
    entry: Optional[BookEntry]
    in_book: bool


@dataclass
class GameForBook:
    """Game result for book generation."""
    moves: list[Move]
    result: Literal[1, 0, -1]  # 1=white win, 0=draw, -1=black win
    evaluations: Optional[list[float]] = None


@dataclass
class BookGenerationOptions:
    """Options for book generation."""
    max_depth: int = 20
    min_position_count: int = 2


# ============================================================================
# Global Opening Book Instance
# ============================================================================

_opening_book: OpeningBook = OpeningBook(
    metadata=BookMetadata(created_at=datetime.now().isoformat())
)


def get_opening_book() -> OpeningBook:
    """Get the current opening book."""
    return _opening_book


def set_opening_book(book: OpeningBook) -> None:
    """Set/replace the opening book."""
    global _opening_book
    _opening_book = book


def clear_opening_book() -> None:
    """Clear the opening book."""
    global _opening_book
    _opening_book = OpeningBook(
        metadata=BookMetadata(created_at=datetime.now().isoformat())
    )


def get_book_entry(board: BoardState) -> Optional[BookEntry]:
    """Get book entry for a position."""
    hash_key = generate_board_hash(board)
    return _opening_book.entries.get(hash_key)


def is_in_book(board: BoardState) -> bool:
    """Check if a position is in the opening book."""
    return get_book_entry(board) is not None


def get_book_size() -> int:
    """Get the number of positions in the book."""
    return len(_opening_book.entries)


# ============================================================================
# Move Selection
# ============================================================================


def create_seeded_random(seed: int) -> random.Random:
    """Create a seeded random generator."""
    rng = random.Random()
    rng.seed(seed)
    return rng


def calculate_win_rate(stats: BookMoveStats) -> float:
    """
    Calculate win rate for a book move.
    Returns value between 0 and 1.
    """
    losses = stats.play_count - stats.wins - stats.draws
    total = stats.wins + stats.draws + losses
    if total == 0:
        return 0.5  # No data, assume neutral
    # Draws count as half a win
    return (stats.wins + stats.draws * 0.5) / total


def calculate_move_weight(
    stats: BookMoveStats,
    options: BookLookupOptions
) -> float:
    """
    Calculate selection weight for a book move.
    Combines play count and win rate.
    """
    temperature = options.temperature
    
    # Base weight from play count (popularity)
    weight = math.sqrt(stats.play_count)
    
    # Adjust by win rate if enabled
    if options.use_win_rate_weight:
        win_rate = calculate_win_rate(stats)
        # Scale win rate to 0.5-1.5 range to adjust weight
        win_rate_multiplier = 0.5 + win_rate
        weight *= win_rate_multiplier
    
    # Apply temperature (higher temperature = more randomness)
    if temperature > 0:
        weight = math.pow(weight, 1.0 / temperature)
    
    return weight


def select_book_move(
    moves: list[BookMoveStats],
    options: BookLookupOptions,
    rng: random.Random
) -> Optional[BookMoveStats]:
    """Select a move from book moves probabilistically."""
    min_play_count = options.min_play_count
    
    # Filter moves by minimum play count
    eligible_moves = [m for m in moves if m.play_count >= min_play_count]
    
    if not eligible_moves:
        return None
    
    # Calculate weights for all eligible moves
    weights = [calculate_move_weight(m, options) for m in eligible_moves]
    total_weight = sum(weights)
    
    if total_weight == 0:
        return None
    
    # Weighted random selection
    r = rng.random() * total_weight
    for i, move in enumerate(eligible_moves):
        r -= weights[i]
        if r <= 0:
            return move
    
    # Fallback to last move (shouldn't happen normally)
    return eligible_moves[-1]


def find_matching_move(
    legal_moves: list[Move],
    book_stats: BookMoveStats
) -> Optional[Move]:
    """Find a move from the book entry that matches the given coordinates."""
    for move in legal_moves:
        if (
            move.from_coord.q == book_stats.from_q and
            move.from_coord.r == book_stats.from_r and
            move.to_coord.q == book_stats.to_q and
            move.to_coord.r == book_stats.to_r and
            move.promotion == book_stats.promotion
        ):
            return move
    return None


def lookup_book_move(
    board: BoardState,
    color: Color,
    options: Optional[BookLookupOptions] = None
) -> BookLookupResult:
    """
    Lookup a move from the opening book for the given position.
    
    Args:
        board: The current board position
        color: The side to move
        options: Lookup options for move selection
        
    Returns:
        BookLookupResult with selected move (or None if not in book)
    """
    if options is None:
        options = BookLookupOptions()
    
    entry = get_book_entry(board)
    
    if entry is None or not entry.moves:
        return BookLookupResult(move=None, entry=None, in_book=False)
    
    # Create RNG for selection
    seed = options.seed if options.seed is not None else int(datetime.now().timestamp() * 1000)
    rng = create_seeded_random(seed)
    
    # Select a book move
    selected_stats = select_book_move(entry.moves, options, rng)
    
    if selected_stats is None:
        return BookLookupResult(move=None, entry=entry, in_book=True)
    
    # Get legal moves and find the matching one
    legal_moves = generate_all_legal_moves(board, color)
    move = find_matching_move(legal_moves, selected_stats)
    
    return BookLookupResult(move=move, entry=entry, in_book=True)


# ============================================================================
# Book Generation from Self-Play
# ============================================================================


def add_game_to_book(
    game: GameForBook,
    starting_board: BoardState,
    options: Optional[BookGenerationOptions] = None
) -> None:
    """
    Add a single game to the opening book.
    
    Args:
        game: The game to add
        starting_board: The starting position (for replay)
        options: Generation options
    """
    if options is None:
        options = BookGenerationOptions()
    
    max_depth = options.max_depth
    
    current_board = dict(starting_board)
    current_color: Color = "white"
    
    for ply in range(min(len(game.moves), max_depth)):
        move = game.moves[ply]
        hash_key = generate_board_hash(current_board)
        
        # Get or create book entry for this position
        entry = _opening_book.entries.get(hash_key)
        if entry is None:
            entry = BookEntry(hash_key=hash_key)
            _opening_book.entries[hash_key] = entry
        
        entry.total_visits += 1
        
        # Find or create move stats
        move_stats: Optional[BookMoveStats] = None
        for m in entry.moves:
            if (
                m.from_q == move.from_coord.q and
                m.from_r == move.from_coord.r and
                m.to_q == move.to_coord.q and
                m.to_r == move.to_coord.r and
                m.promotion == move.promotion
            ):
                move_stats = m
                break
        
        if move_stats is None:
            move_stats = BookMoveStats(
                from_q=move.from_coord.q,
                from_r=move.from_coord.r,
                to_q=move.to_coord.q,
                to_r=move.to_coord.r,
                promotion=move.promotion,
            )
            entry.moves.append(move_stats)
        
        move_stats.play_count += 1
        
        # Update win/draw stats based on game result
        # Win for the side that played the move
        if game.result == 1 and current_color == "white":
            move_stats.wins += 1
        elif game.result == -1 and current_color == "black":
            move_stats.wins += 1
        elif game.result == 0:
            move_stats.draws += 1
        
        # Update average score if evaluations provided
        if game.evaluations and ply < len(game.evaluations):
            eval_score = game.evaluations[ply]
            # Convert to moving side's perspective
            side_eval = eval_score if current_color == "white" else -eval_score
            # Running average
            move_stats.avg_score = (
                (move_stats.avg_score * (move_stats.play_count - 1) + side_eval) /
                move_stats.play_count
            )
        
        # Apply the move
        current_board = apply_move(current_board, move)
        current_color = opposite_color(current_color)
        
        # Update max depth in metadata
        if ply + 1 > _opening_book.metadata.max_depth:
            _opening_book.metadata.max_depth = ply + 1
    
    _opening_book.metadata.games_count += 1


def generate_opening_book(
    games: list[GameForBook],
    starting_board: BoardState,
    options: Optional[BookGenerationOptions] = None
) -> OpeningBook:
    """
    Generate an opening book from multiple games.
    
    Args:
        games: Array of games to include
        starting_board: The starting position
        options: Generation options
        
    Returns:
        The generated opening book
    """
    if options is None:
        options = BookGenerationOptions()
    
    # Clear and reset
    clear_opening_book()
    
    # Add all games
    for game in games:
        add_game_to_book(game, starting_board, options)
    
    # Prune positions with too few visits
    prune_book(options.min_position_count)
    
    _opening_book.metadata.created_at = datetime.now().isoformat()
    
    return _opening_book


def prune_book(min_count: int) -> None:
    """Remove positions with fewer than min_count visits."""
    to_delete: list[str] = []
    
    for hash_key, entry in _opening_book.entries.items():
        if entry.total_visits < min_count:
            to_delete.append(hash_key)
        else:
            # Also prune moves with too few plays within each entry
            min_move_count = max(1, min_count // 2)
            entry.moves = [m for m in entry.moves if m.play_count >= min_move_count]
    
    for hash_key in to_delete:
        del _opening_book.entries[hash_key]


# ============================================================================
# Book Serialization
# ============================================================================


def book_move_stats_to_dict(stats: BookMoveStats) -> dict[str, Any]:
    """Convert BookMoveStats to dictionary for JSON serialization."""
    d: dict[str, Any] = {
        "from_q": stats.from_q,
        "from_r": stats.from_r,
        "to_q": stats.to_q,
        "to_r": stats.to_r,
        "play_count": stats.play_count,
        "wins": stats.wins,
        "draws": stats.draws,
        "avg_score": stats.avg_score,
    }
    if stats.promotion is not None:
        d["promotion"] = stats.promotion
    return d


def dict_to_book_move_stats(d: dict[str, Any]) -> BookMoveStats:
    """Convert dictionary to BookMoveStats."""
    return BookMoveStats(
        from_q=d["from_q"],
        from_r=d["from_r"],
        to_q=d["to_q"],
        to_r=d["to_r"],
        play_count=d.get("play_count", 0),
        wins=d.get("wins", 0),
        draws=d.get("draws", 0),
        avg_score=d.get("avg_score", 0.0),
        promotion=d.get("promotion"),
    )


def serialize_book(book: Optional[OpeningBook] = None) -> dict[str, Any]:
    """Serialize opening book to JSON-compatible format."""
    if book is None:
        book = _opening_book
    
    entries: list[dict[str, Any]] = []
    
    for hash_key, entry in book.entries.items():
        entries.append({
            "hash": hash_key,
            "moves": [book_move_stats_to_dict(m) for m in entry.moves],
            "totalVisits": entry.total_visits,
        })
    
    return {
        "entries": entries,
        "metadata": {
            "createdAt": book.metadata.created_at,
            "gamesCount": book.metadata.games_count,
            "maxDepth": book.metadata.max_depth,
        },
    }


def deserialize_book(data: dict[str, Any]) -> OpeningBook:
    """Deserialize opening book from JSON format."""
    entries: dict[str, BookEntry] = {}
    
    for entry_data in data.get("entries", []):
        hash_key = entry_data["hash"]
        moves = [dict_to_book_move_stats(m) for m in entry_data.get("moves", [])]
        entries[hash_key] = BookEntry(
            hash_key=hash_key,
            moves=moves,
            total_visits=entry_data.get("totalVisits", 0),
        )
    
    metadata_data = data.get("metadata", {})
    metadata = BookMetadata(
        created_at=metadata_data.get("createdAt", ""),
        games_count=metadata_data.get("gamesCount", 0),
        max_depth=metadata_data.get("maxDepth", 0),
    )
    
    return OpeningBook(entries=entries, metadata=metadata)


def export_book_to_json(book: Optional[OpeningBook] = None) -> str:
    """Export opening book to JSON string."""
    return json.dumps(serialize_book(book), indent=2)


def import_book_from_json(json_str: str) -> OpeningBook:
    """Import opening book from JSON string."""
    data = json.loads(json_str)
    return deserialize_book(data)


def load_book_from_json(json_str: str) -> None:
    """Load opening book from JSON string and set as active book."""
    book = import_book_from_json(json_str)
    set_opening_book(book)


def save_book_to_file(filepath: str, book: Optional[OpeningBook] = None) -> None:
    """Save opening book to a JSON file."""
    json_str = export_book_to_json(book)
    with open(filepath, "w", encoding="utf-8") as f:
        f.write(json_str)


def load_book_from_file(filepath: str) -> OpeningBook:
    """Load opening book from a JSON file."""
    with open(filepath, "r", encoding="utf-8") as f:
        json_str = f.read()
    return import_book_from_json(json_str)


# ============================================================================
# Book Statistics
# ============================================================================


@dataclass
class BookStatistics:
    """Statistics about the opening book."""
    position_count: int = 0
    total_move_entries: int = 0
    avg_moves_per_position: float = 0.0
    max_depth: int = 0
    games_count: int = 0
    most_visited_position: Optional[str] = None
    highest_play_count: int = 0


def get_book_statistics() -> BookStatistics:
    """Get statistics about the current opening book."""
    total_move_entries = 0
    most_visited_position: Optional[str] = None
    most_visits = 0
    highest_play_count = 0
    
    for hash_key, entry in _opening_book.entries.items():
        total_move_entries += len(entry.moves)
        
        if entry.total_visits > most_visits:
            most_visits = entry.total_visits
            most_visited_position = hash_key
        
        for move in entry.moves:
            if move.play_count > highest_play_count:
                highest_play_count = move.play_count
    
    position_count = len(_opening_book.entries)
    
    return BookStatistics(
        position_count=position_count,
        total_move_entries=total_move_entries,
        avg_moves_per_position=(
            total_move_entries / position_count if position_count > 0 else 0.0
        ),
        max_depth=_opening_book.metadata.max_depth,
        games_count=_opening_book.metadata.games_count,
        most_visited_position=most_visited_position,
        highest_play_count=highest_play_count,
    )


def format_book_entry(entry: BookEntry) -> str:
    """Format book entry for display (debugging/UI)."""
    lines: list[str] = []
    lines.append(f"Position hash: {entry.hash_key} (visited {entry.total_visits} times)")
    lines.append("Moves:")
    
    # Sort by play count
    sorted_moves = sorted(entry.moves, key=lambda m: m.play_count, reverse=True)
    
    for move in sorted_moves:
        win_rate = calculate_win_rate(move)
        from_str = f"{move.from_q},{move.from_r}"
        to_str = f"{move.to_q},{move.to_r}"
        promo = f" (={move.promotion})" if move.promotion else ""
        lines.append(
            f"  {from_str}-{to_str}{promo}: "
            f"played={move.play_count}, wins={move.wins}, draws={move.draws}, "
            f"winRate={win_rate * 100:.1f}%"
        )
    
    return "\n".join(lines)

File: src/typescript/src/openingbook.ts
/**
 * Underchex Opening Book Module
 * 
 * Provides opening book functionality for the AI:
 * - Store and lookup opening book positions
 * - Generate opening book from self-play games
 * - Probabilistic move selection based on win rates
 * 
 * Opening book format:
 * - Positions are keyed by Zobrist hash
 * - Each position stores a list of candidate moves with statistics
 * - Moves are selected probabilistically based on win rate and play count
 * 
 * Signed-by: agent #27 claude-sonnet-4 via opencode 20260122T07:49:00
 */

import {
  BoardState,
  Move,
  Color,
  HexCoord,
  PieceType,
  coordToString,
  oppositeColor,
} from './types';

import {
  computeZobristHash,
  PIECE_VALUES,
} from './ai';

import {
  generateAllLegalMoves,
  applyMove,
} from './moves';

import {
  getStartingPosition as getStartingPlacements,
  createBoardFromPlacements,
} from './game';

// ============================================================================
// Opening Book Types
// ============================================================================

/**
 * Statistics for a single move in the opening book.
 */
export interface BookMoveStats {
  /** Move from position */
  from: HexCoord;
  /** Move to position */
  to: HexCoord;
  /** Number of times this move was played */
  playCount: number;
  /** Number of times this move led to a win for the side that played it */
  wins: number;
  /** Number of draws when this move was played */
  draws: number;
  /** Average score from positions after this move (from the moving side's perspective) */
  avgScore: number;
  /** Promotion piece type if this is a promotion move */
  promotion?: PieceType;
}

/**
 * Entry in the opening book for a single position.
 */
export interface BookEntry {
  /** Zobrist hash of the position */
  hash: number;
  /** Candidate moves with their statistics */
  moves: BookMoveStats[];
  /** Total times this position was reached */
  totalVisits: number;
}

/**
 * The full opening book structure.
 */
export interface OpeningBook {
  /** Map from Zobrist hash to book entry */
  entries: Map<number, BookEntry>;
  /** Metadata */
  metadata: {
    /** When the book was created/updated */
    createdAt: string;
    /** Number of games used to generate the book */
    gamesCount: number;
    /** Maximum depth (ply) of positions in the book */
    maxDepth: number;
  };
}

/**
 * Options for opening book lookup.
 */
export interface BookLookupOptions {
  /** Minimum play count for a move to be considered (default: 3) */
  minPlayCount?: number;
  /** Temperature for probabilistic selection (0 = always best, higher = more random) */
  temperature?: number;
  /** Whether to use win rate weighting (default: true) */
  useWinRateWeight?: boolean;
  /** Random seed for deterministic selection (optional) */
  seed?: number;
}

/**
 * Result of opening book lookup.
 */
export interface BookLookupResult {
  /** Selected move, or null if no book move available */
  move: Move | null;
  /** Book entry for the position (for debugging/UI) */
  entry: BookEntry | null;
  /** Whether the move was found in the book */
  inBook: boolean;
}

// ============================================================================
// Opening Book Implementation
// ============================================================================

/**
 * Default opening book (starts empty, populated by generateBook or loadBook).
 */
let openingBook: OpeningBook = {
  entries: new Map(),
  metadata: {
    createdAt: new Date().toISOString(),
    gamesCount: 0,
    maxDepth: 0,
  },
};

/**
 * Get the current opening book.
 */
export function getOpeningBook(): OpeningBook {
  return openingBook;
}

/**
 * Set/replace the opening book.
 */
export function setOpeningBook(book: OpeningBook): void {
  openingBook = book;
}

/**
 * Clear the opening book.
 */
export function clearOpeningBook(): void {
  openingBook = {
    entries: new Map(),
    metadata: {
      createdAt: new Date().toISOString(),
      gamesCount: 0,
      maxDepth: 0,
    },
  };
}

/**
 * Get book entry for a position.
 */
export function getBookEntry(board: BoardState): BookEntry | null {
  const hash = computeZobristHash(board);
  return openingBook.entries.get(hash) ?? null;
}

/**
 * Check if a position is in the opening book.
 */
export function isInBook(board: BoardState): boolean {
  return getBookEntry(board) !== null;
}

/**
 * Get the number of positions in the book.
 */
export function getBookSize(): number {
  return openingBook.entries.size;
}

// ============================================================================
// Move Selection
// ============================================================================

/**
 * Simple seeded PRNG for deterministic random selection.
 * Uses xorshift32 algorithm.
 */
function createSeededRandom(seed: number): () => number {
  let state = seed;
  return () => {
    state ^= state << 13;
    state ^= state >>> 17;
    state ^= state << 5;
    return (state >>> 0) / 0xFFFFFFFF; // Returns 0-1
  };
}

/**
 * Calculate win rate for a book move.
 * Returns value between 0 and 1.
 */
export function calculateWinRate(stats: BookMoveStats): number {
  const total = stats.wins + stats.draws + (stats.playCount - stats.wins - stats.draws);
  if (total === 0) return 0.5; // No data, assume neutral
  // Draws count as half a win
  return (stats.wins + stats.draws * 0.5) / total;
}

/**
 * Calculate selection weight for a book move.
 * Combines play count and win rate.
 */
function calculateMoveWeight(
  stats: BookMoveStats,
  options: BookLookupOptions
): number {
  const useWinRate = options.useWinRateWeight ?? true;
  const temperature = options.temperature ?? 1.0;
  
  // Base weight from play count (popularity)
  let weight = Math.sqrt(stats.playCount);
  
  // Adjust by win rate if enabled
  if (useWinRate) {
    const winRate = calculateWinRate(stats);
    // Scale win rate to 0.5-1.5 range to adjust weight
    const winRateMultiplier = 0.5 + winRate;
    weight *= winRateMultiplier;
  }
  
  // Apply temperature (higher temperature = more randomness)
  if (temperature > 0) {
    weight = Math.pow(weight, 1.0 / temperature);
  }
  
  return weight;
}

/**
 * Select a move from book moves probabilistically.
 */
function selectBookMove(
  moves: BookMoveStats[],
  options: BookLookupOptions,
  rng: () => number
): BookMoveStats | null {
  const minPlayCount = options.minPlayCount ?? 3;
  
  // Filter moves by minimum play count
  const eligibleMoves = moves.filter(m => m.playCount >= minPlayCount);
  
  if (eligibleMoves.length === 0) {
    return null;
  }
  
  // Calculate weights for all eligible moves
  const weights = eligibleMoves.map(m => calculateMoveWeight(m, options));
  const totalWeight = weights.reduce((a, b) => a + b, 0);
  
  if (totalWeight === 0) {
    return null;
  }
  
  // Weighted random selection
  let random = rng() * totalWeight;
  for (let i = 0; i < eligibleMoves.length; i++) {
    random -= weights[i]!;
    if (random <= 0) {
      return eligibleMoves[i]!;
    }
  }
  
  // Fallback to last move (shouldn't happen normally)
  return eligibleMoves[eligibleMoves.length - 1]!;
}

/**
 * Find a move from the book entry that matches the given coordinates.
 */
function findMatchingMove(
  legalMoves: Move[],
  bookStats: BookMoveStats
): Move | null {
  for (const move of legalMoves) {
    if (
      move.from.q === bookStats.from.q &&
      move.from.r === bookStats.from.r &&
      move.to.q === bookStats.to.q &&
      move.to.r === bookStats.to.r &&
      move.promotion === bookStats.promotion
    ) {
      return move;
    }
  }
  return null;
}

/**
 * Lookup a move from the opening book for the given position.
 * 
 * @param board The current board position
 * @param color The side to move
 * @param options Lookup options for move selection
 * @returns BookLookupResult with selected move (or null if not in book)
 */
export function lookupBookMove(
  board: BoardState,
  color: Color,
  options: BookLookupOptions = {}
): BookLookupResult {
  const entry = getBookEntry(board);
  
  if (!entry || entry.moves.length === 0) {
    return { move: null, entry: null, inBook: false };
  }
  
  // Create RNG for selection
  const seed = options.seed ?? Date.now();
  const rng = createSeededRandom(seed);
  
  // Select a book move
  const selectedStats = selectBookMove(entry.moves, options, rng);
  
  if (!selectedStats) {
    return { move: null, entry, inBook: true };
  }
  
  // Get legal moves and find the matching one
  const legalMoves = generateAllLegalMoves(board, color);
  const move = findMatchingMove(legalMoves, selectedStats);
  
  return { move, entry, inBook: true };
}

// ============================================================================
// Book Generation from Self-Play
// ============================================================================

/**
 * Game result for book generation.
 */
export interface GameForBook {
  /** Sequence of moves played */
  moves: Move[];
  /** Game result: 1 = white win, 0 = draw, -1 = black win */
  result: 1 | 0 | -1;
  /** Optional: position evaluations after each move */
  evaluations?: number[];
}

/**
 * Options for book generation.
 */
export interface BookGenerationOptions {
  /** Maximum depth (ply) to include in the book (default: 20) */
  maxDepth?: number;
  /** Minimum times a position must be seen to be included (default: 2) */
  minPositionCount?: number;
}

/**
 * Add a single game to the opening book.
 * 
 * @param game The game to add
 * @param startingBoard The starting position (for replay)
 * @param options Generation options
 */
export function addGameToBook(
  game: GameForBook,
  startingBoard: BoardState,
  options: BookGenerationOptions = {}
): void {
  const maxDepth = options.maxDepth ?? 20;
  
  let currentBoard = new Map(startingBoard);
  let currentColor: Color = 'white';
  
  for (let ply = 0; ply < Math.min(game.moves.length, maxDepth); ply++) {
    const move = game.moves[ply]!;
    const hash = computeZobristHash(currentBoard);
    
    // Get or create book entry for this position
    let entry = openingBook.entries.get(hash);
    if (!entry) {
      entry = {
        hash,
        moves: [],
        totalVisits: 0,
      };
      openingBook.entries.set(hash, entry);
    }
    
    entry.totalVisits++;
    
    // Find or create move stats
    let moveStats = entry.moves.find(
      m => m.from.q === move.from.q &&
           m.from.r === move.from.r &&
           m.to.q === move.to.q &&
           m.to.r === move.to.r &&
           m.promotion === move.promotion
    );
    
    if (!moveStats) {
      moveStats = {
        from: { q: move.from.q, r: move.from.r },
        to: { q: move.to.q, r: move.to.r },
        playCount: 0,
        wins: 0,
        draws: 0,
        avgScore: 0,
        promotion: move.promotion,
      };
      entry.moves.push(moveStats);
    }
    
    moveStats.playCount++;
    
    // Update win/draw stats based on game result
    // Win for the side that played the move
    if (game.result === 1 && currentColor === 'white') {
      moveStats.wins++;
    } else if (game.result === -1 && currentColor === 'black') {
      moveStats.wins++;
    } else if (game.result === 0) {
      moveStats.draws++;
    }
    
    // Update average score if evaluations provided
    if (game.evaluations && game.evaluations[ply] !== undefined) {
      const eval_ = game.evaluations[ply]!;
      // Convert to moving side's perspective
      const sideEval = currentColor === 'white' ? eval_ : -eval_;
      // Running average
      moveStats.avgScore = 
        (moveStats.avgScore * (moveStats.playCount - 1) + sideEval) / moveStats.playCount;
    }
    
    // Apply the move
    currentBoard = applyMove(currentBoard, move);
    currentColor = oppositeColor(currentColor);
    
    // Update max depth in metadata
    if (ply + 1 > openingBook.metadata.maxDepth) {
      openingBook.metadata.maxDepth = ply + 1;
    }
  }
  
  openingBook.metadata.gamesCount++;
}

/**
 * Generate an opening book from multiple games.
 * 
 * @param games Array of games to include
 * @param startingBoard The starting position
 * @param options Generation options
 * @returns The generated opening book
 */
export function generateOpeningBook(
  games: GameForBook[],
  startingBoard: BoardState,
  options: BookGenerationOptions = {}
): OpeningBook {
  // Clear and reset
  clearOpeningBook();
  
  // Add all games
  for (const game of games) {
    addGameToBook(game, startingBoard, options);
  }
  
  // Prune positions with too few visits
  const minCount = options.minPositionCount ?? 2;
  pruneBook(minCount);
  
  openingBook.metadata.createdAt = new Date().toISOString();
  
  return openingBook;
}

/**
 * Remove positions with fewer than minCount visits.
 */
export function pruneBook(minCount: number): void {
  const entries = openingBook.entries;
  const toDelete: number[] = [];
  
  for (const [hash, entry] of entries) {
    if (entry.totalVisits < minCount) {
      toDelete.push(hash);
    } else {
      // Also prune moves with too few plays within each entry
      entry.moves = entry.moves.filter(m => m.playCount >= Math.max(1, Math.floor(minCount / 2)));
    }
  }
  
  for (const hash of toDelete) {
    entries.delete(hash);
  }
}

// ============================================================================
// Book Serialization
// ============================================================================

/**
 * Serialized format for book entry (JSON-compatible).
 */
interface SerializedBookEntry {
  hash: number;
  moves: BookMoveStats[];
  totalVisits: number;
}

/**
 * Serialized format for opening book (JSON-compatible).
 */
export interface SerializedOpeningBook {
  entries: SerializedBookEntry[];
  metadata: OpeningBook['metadata'];
}

/**
 * Serialize opening book to JSON-compatible format.
 */
export function serializeBook(book: OpeningBook = openingBook): SerializedOpeningBook {
  const entries: SerializedBookEntry[] = [];
  
  for (const [hash, entry] of book.entries) {
    entries.push({
      hash,
      moves: entry.moves,
      totalVisits: entry.totalVisits,
    });
  }
  
  return {
    entries,
    metadata: book.metadata,
  };
}

/**
 * Deserialize opening book from JSON format.
 */
export function deserializeBook(data: SerializedOpeningBook): OpeningBook {
  const entries = new Map<number, BookEntry>();
  
  for (const entry of data.entries) {
    entries.set(entry.hash, {
      hash: entry.hash,
      moves: entry.moves,
      totalVisits: entry.totalVisits,
    });
  }
  
  return {
    entries,
    metadata: data.metadata,
  };
}

/**
 * Export opening book to JSON string.
 */
export function exportBookToJSON(book: OpeningBook = openingBook): string {
  return JSON.stringify(serializeBook(book), null, 2);
}

/**
 * Import opening book from JSON string.
 */
export function importBookFromJSON(json: string): OpeningBook {
  const data = JSON.parse(json) as SerializedOpeningBook;
  return deserializeBook(data);
}

/**
 * Load opening book from JSON string and set as active book.
 */
export function loadBookFromJSON(json: string): void {
  const book = importBookFromJSON(json);
  setOpeningBook(book);
}

// ============================================================================
// Book Statistics
// ============================================================================

/**
 * Statistics about the opening book.
 */
export interface BookStatistics {
  /** Total number of positions */
  positionCount: number;
  /** Total number of move entries (across all positions) */
  totalMoveEntries: number;
  /** Average moves per position */
  avgMovesPerPosition: number;
  /** Maximum depth in the book */
  maxDepth: number;
  /** Number of games used to generate */
  gamesCount: number;
  /** Most visited position hash */
  mostVisitedPosition: number | null;
  /** Highest play count for any single move */
  highestPlayCount: number;
}

/**
 * Get statistics about the current opening book.
 */
export function getBookStatistics(): BookStatistics {
  let totalMoveEntries = 0;
  let mostVisitedPosition: number | null = null;
  let mostVisits = 0;
  let highestPlayCount = 0;
  
  for (const [hash, entry] of openingBook.entries) {
    totalMoveEntries += entry.moves.length;
    
    if (entry.totalVisits > mostVisits) {
      mostVisits = entry.totalVisits;
      mostVisitedPosition = hash;
    }
    
    for (const move of entry.moves) {
      if (move.playCount > highestPlayCount) {
        highestPlayCount = move.playCount;
      }
    }
  }
  
  const positionCount = openingBook.entries.size;
  
  return {
    positionCount,
    totalMoveEntries,
    avgMovesPerPosition: positionCount > 0 ? totalMoveEntries / positionCount : 0,
    maxDepth: openingBook.metadata.maxDepth,
    gamesCount: openingBook.metadata.gamesCount,
    mostVisitedPosition,
    highestPlayCount,
  };
}

/**
 * Format book entry for display (debugging/UI).
 */
export function formatBookEntry(entry: BookEntry): string {
  const lines: string[] = [];
  lines.push(`Position hash: ${entry.hash} (visited ${entry.totalVisits} times)`);
  lines.push('Moves:');
  
  // Sort by play count
  const sortedMoves = [...entry.moves].sort((a, b) => b.playCount - a.playCount);
  
  for (const move of sortedMoves) {
    const winRate = calculateWinRate(move);
    const fromStr = coordToString(move.from);
    const toStr = coordToString(move.to);
    const promo = move.promotion ? ` (=${move.promotion})` : '';
    lines.push(
      `  ${fromStr}-${toStr}${promo}: ` +
      `played=${move.playCount}, wins=${move.wins}, draws=${move.draws}, ` +
      `winRate=${(winRate * 100).toFixed(1)}%`
    );
  }
  
  return lines.join('\n');
}
